rclcpp_action/src/client.cpp

```diff
@@ -12,30 +12,349 @@
+{
+  GoalID goal_id;
+  // TODO(hidmic): Do something better than this for UUID generation.
+  // std::generate(
+  //   goal_id.uuid.begin(), goal_id.uuid.end(),
+  //   std::ref(pimpl_->random_bytes_generator));
+  std::generate(
+    goal_id.begin(), goal_id.end(),
+    std::ref(pimpl_->random_bytes_generator));
+  return goal_id;
+}
+
+void
+ClientBase::execute()
+{
+  if (pimpl_->is_feedback_ready) {
+    std::shared_ptr<void> feedback_message = this->create_feedback_message();
+    rcl_ret_t ret = rcl_action_take_feedback(
+      pimpl_->client_handle.get(), feedback_message.get());
+    pimpl_->is_feedback_ready = false;
+    if (RCL_RET_OK != ret) {
+      rclcpp::exceptions::throw_from_rcl_error(ret, "error taking feedback");
+    } else {
+      this->handle_feedback_message(feedback_message);
+    }
+  } else if (pimpl_->is_status_ready) {
+    std::shared_ptr<void> status_message = this->create_status_message();
+    rcl_ret_t ret = rcl_action_take_status(
+      pimpl_->client_handle.get(), status_message.get());
+    pimpl_->is_status_ready = false;
+    if (RCL_RET_OK != ret) {
+      rclcpp::exceptions::throw_from_rcl_error(ret, "error taking status");
+    } else {
+      this->handle_status_message(status_message);
+    }
+  } else if (pimpl_->is_goal_response_ready) {
+    rmw_request_id_t response_header;
+    std::shared_ptr<void> goal_response = this->create_goal_response();
+    rcl_ret_t ret = rcl_action_take_goal_response(
+      pimpl_->client_handle.get(), &response_header, goal_response.get());
+    pimpl_->is_goal_response_ready = false;
+    if (RCL_RET_OK != ret) {
+      rclcpp::exceptions::throw_from_rcl_error(ret, "error taking goal response");
+    } else {
+      this->handle_goal_response(response_header, goal_response);
+    }
+  } else if (pimpl_->is_result_response_ready) {
+    rmw_request_id_t response_header;
+    std::shared_ptr<void> result_response = this->create_result_response();
+    rcl_ret_t ret = rcl_action_take_result_response(
+      pimpl_->client_handle.get(), &response_header, result_response.get());
+    pimpl_->is_result_response_ready = false;
+    if (RCL_RET_OK != ret) {
+      rclcpp::exceptions::throw_from_rcl_error(ret, "error taking result response");
+    } else {
+      this->handle_result_response(response_header, result_response);
+    }
+  } else if (pimpl_->is_cancel_response_ready) {
+    rmw_request_id_t response_header;
+    std::shared_ptr<void> cancel_response = this->create_cancel_response();
+    rcl_ret_t ret = rcl_action_take_cancel_response(
+      pimpl_->client_handle.get(), &response_header, cancel_response.get());
+    pimpl_->is_cancel_response_ready = false;
+    if (RCL_RET_OK != ret) {
+      rclcpp::exceptions::throw_from_rcl_error(ret, "error taking cancel response");
+    } else {
+      this->handle_cancel_response(response_header, cancel_response);
+    }
+  } else {
+    throw std::runtime_error("Executing action client but nothing is ready");
```

Comment:
@hidmic, can I ask what was the purpose of throwing this exception? It's throwing in my code and as far as I can tell there's nothing I can do to control it. And since this function is invoked through the executor, I can't easily catch it and proceed. When I remove it, my program functions normally. Is it just to signal that there's a bug? I

Reply:
> Is it just to signal that there's a bug?

I think so. Got an example that reproduces it? It would mean the executor thinks the action client is ready, but nothing actually seems to be ready to execute.

Reply:
I'll see if I can get a simple standalone example reproducing it since my current program is pretty complex. I'll just go ahead an open an issue in that case.

Reply:
> It would mean the executor thinks the action client is ready, but nothing actually seems to be ready to execute.

Precisely that @jdlangs.